File: 52-n_queens_II.js
/**
 * Solves the N-Queens II problem by counting the number of valid placements.
 * @param {number} n - The size of the chessboard (n x n).
 * @return {number} - The number of valid ways to place N queens.
 */
var totalNQueens = function(n) {
    let count = 0; // Stores the total number of valid solutions
    
    // Sets to track columns and diagonals under attack
    let cols = new Set();         // Tracks occupied columns
    let mainDiag = new Set();      // Tracks main diagonals (row - col)
    let antiDiag = new Set();      // Tracks anti-diagonals (row + col)

    /**
     * Backtracking function to place queens row by row
     * @param {number} row - The current row being considered.
     */
    function backtrack(row) {
        if (row === n) {
            // If we reach row 'n', we have successfully placed all queens
            count++;
            return;
        }

        for (let col = 0; col < n; col++) {
            // Check if the current column or diagonals are under attack
            if (cols.has(col) || mainDiag.has(row - col) || antiDiag.has(row + col)) {
                continue; // Skip this position if it's not safe
            }

            // Place the queen
            cols.add(col);
            mainDiag.add(row - col);
            antiDiag.add(row + col);

            // Recurse to the next row
            backtrack(row + 1);

            // Remove the queen (backtrack)
            cols.delete(col);
            mainDiag.delete(row - col);
            antiDiag.delete(row + col);
        }
    }

    // Start backtracking from the first row
    backtrack(0);
    
    return count; // Return the total number of solutions
};

// Example Usage
console.log(totalNQueens(4)); // Output: 2
console.log(totalNQueens(8)); // Output: 92

File: 52-n_queens_II.py
class Solution:
    def totalNQueens(self, n: int) -> int:
        # This function will use backtracking to count the number of valid solutions.
        def backtrack(row: int):
            if row == n:
                # If we have placed queens in all rows, we found a valid solution
                nonlocal count
                count += 1
                return
            
            for col in range(n):
                # Check if the current position is under attack
                if col in cols or (row - col) in diagonals or (row + col) in anti_diagonals:
                    continue  # Skip this position if it's under attack
                
                # Place the queen
                cols.add(col)
                diagonals.add(row - col)
                anti_diagonals.add(row + col)
                
                # Move to the next row
                backtrack(row + 1)
                
                # Backtrack: remove the queen and try next column
                cols.remove(col)
                diagonals.remove(row - col)
                anti_diagonals.remove(row + col)
        
        count = 0  # To store the total number of valid solutions
        cols = set()  # Columns where queens are placed
        diagonals = set()  # Difference of row and column (\ diagonal)
        anti_diagonals = set()  # Sum of row and column (/ diagonal)
        
        backtrack(0)  # Start from the first row
        return count
